new-package-set.nix

```diff
@@ -0,0 +1,187 @@
+                  else rev;
+              };
+              contentAddressedRevs = lib.foldr f {} (builtins.attrValues version.revisions);
+            in lib.mapAttrs (_: rev2Config) (version.revisions // contentAddressedRevs);
+        }));
+
+  # The planFunc (that is, a package set description like an LTS set
+  # or a plan.nix (derived from plan.json)) will then producde a structure
+  # that looks like:
+  #
+  # { packages = { "package" = { revision = hackageConfigs.$package.$version.revisions.default;
+  #                              flags = { flag1 = true; flag2 = false; ... }; };
+  #                ... };
+  #   compiler = { version = "X.Y.Z"; nix-name ="ghcXYZ";
+  #                # packages that come bundled with the compiler
+  #                packages = { "bytestring" = "a.b.c.d"; ... }; };
+  # }
+
+  # package descriptions in hackage will look like:
+  # { system, compiler, flags, pkgs, hsPkgs, pkgconfPkgs }:
+  # { flags = { flag1 = false; flags2 = true; ... };
+  #   package = { specVersion = "X.Y"; identifier = { name = "..."; version = "a.b.c.d"; };
+  #               license = "..."; copyright = "..."; maintainer = "..."; author = "...";
+  #               homepage = "..."; url = "..."; synopsis = "..."; description = "...";
+  #               buildType = "Simple"; # or Custom, Autoconf, ...
+  #             };
+  #  components = {
+  #    "..." = { depends = [ (hsPkgs.base) ... ]; };
+  #    exes = { "..." = { depends = ... };
+  #             "..." = { depends = ... }; };
+  #    tests = { "..." = { depends = ... }; ... };
+  #  };
+
+  # The package descriptions depend on pkgs, which are used to resolve system package dependencies
+  # as well as pkgconfPkgs, which are used to resolve pkgconfig name to nixpkgs names.  We simply
+  # augment the existing pkgs set with the specific mappings:
+  adjustedPkgs = pkgs // (import ./lib/system-nixpkgs-map.nix pkgs);
+  pkgconfPkgs = pkgs // (import ./lib/pkgconf-nixpkgs-map.nix pkgs);
+
+  # it also crucially depends on system, and compiler, both of which need to be resolved to the
+  # current system being targetted.
+  hostMap = import ./lib/host-map.nix pkgs.stdenv;
+  cabal = import ./lib/cabal-os-arch-comp.nix;
+
+  compiler = cabal.compiler // {
+    isGhc = true;
+    version = lib.mapAttrs (_: f: v: f (builtins.compareVersions plan.compiler.version v)) {
+      eq = c: c == 0;
+      gt = c: c > 0;
+      ge = c: c >= 0;
+      lt = c: c < 0;
+      le = c: c <= 0;
+    };
+  };
+  system = cabal.os   // { "is${hostMap.os}"   = true; }
+        // cabal.arch // { "is${hostMap.arch}" = true; };
+
+  # Therefore the arguments we pass to the package are
+  pkgArgs = { inherit compiler system pkgconfPkgs; pkgs = adjustedPkgs; };
+
+  # we can never reinstall the following set of packages:
+  nonReinstallables
+    = [ "rts" "ghc" "base" "ghc-prim" "array" "integer-gmp" "integer-simple"
+        "deepseq" "pretty" "ghc-boot-th" "template-haskell" "terminfo" ];
```

Comment:
We need `terminfo` here as well, until we can pass flags trivially to packages. The underlying reason is that terminfo needs to be provided `--with-cureses-lirbaries` and `--with-curses-includes` in nix, to record the proper paths: see https://github.com/NixOS/nixpkgs/commit/d1fc5370e93f202c2de5a53ab5620cb29880a103, which does this for the terminfo that's shipped with GHC.

However if we want to build it outside of GHC, we need to pass those flags.